internal/webhook/daemonsetmutation/webhookhandler.go

```diff
@@ -0,0 +1,95 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// Package daemonsetmutation contains the webhook that injects annotations into daemon-sets.
+package daemonsetmutation
+
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+
+	"github.com/go-logr/logr"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	"github.com/aws/amazon-cloudwatch-agent-operator/internal/config"
+)
+
+// +kubebuilder:webhook:path=/mutate-v1-daemonset,mutating=true,failurePolicy=ignore,groups="apps",resources=daemonsets,verbs=create;update,versions=v1,name=mdaemonset.kb.io,sideEffects=none,admissionReviewVersions=v1
+// +kubebuilder:rbac:groups="",resources=namespaces,verbs=list;watch
+// +kubebuilder:rbac:groups="apps",resources=daemonsets,verbs=get;list;watch
+
+var _ WebhookHandler = (*daemonSetMutationWebhook)(nil)
+
+// WebhookHandler is a webhook handler that analyzes new daemon-sets and injects appropriate annotations into it.
+type WebhookHandler interface {
+	admission.Handler
+}
+
+// the implementation.
+type daemonSetMutationWebhook struct {
+	client            client.Client
+	decoder           *admission.Decoder
+	logger            logr.Logger
+	daemonSetMutators []DaemonSetMutator
+	config            config.Config
+}
+
+// DaemonSetMutator mutates a daemon-set.
+type DaemonSetMutator interface {
+	Mutate(ctx context.Context, ds appsv1.DaemonSet) (appsv1.DaemonSet, error)
+}
+
+// NewWebhookHandler creates a new WebhookHandler.
+func NewWebhookHandler(cfg config.Config, logger logr.Logger, decoder *admission.Decoder, cl client.Client, daemonSetMutators []DaemonSetMutator) WebhookHandler {
+	return &daemonSetMutationWebhook{
+		config:            cfg,
+		decoder:           decoder,
+		logger:            logger,
+		client:            cl,
+		daemonSetMutators: daemonSetMutators,
+	}
+}
+
+func (p *daemonSetMutationWebhook) Handle(ctx context.Context, req admission.Request) admission.Response {
+	ds := appsv1.DaemonSet{}
+	err := p.decoder.Decode(req, &ds)
+	if err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	// we use the req.Namespace here because the pod might have not been created yet
+	ns := corev1.Namespace{}
```

Comment:
Dont think we need `ns` here.

Reply:
Ties in to the previous comment, upstream has this change incase the resource is not created yet